### DISCUSSION TITLE
Metadata provider

### DISCUSSION THREAD

**Discussion by zfLQ2qx2** [go-webauthn/webauthn#339] [General]
I don't see any information at all about the Metadata provider. I see that it downloads a blob from the Fido alliance, I found an example of caching that blob with the cache provider and that seems to work, but I can't find any examples of anyone using it. I'd be interested in getting some of the data out like TcDisplayPNGCharacteristics or the PassKeyAuthenticatorAAGUID structure (which looks more like a placeholder and not actually used by the metadata provider) but its not really clear how. Anyone have more information or examples?

**Comment by james-d-elliott** [comment]
I don't have the specific example you're looking for but the concept is relatively straight forward to explain:


```go
import (
       "context"
       "fmt"
       
	"github.com/google/uuid"
	"github.com/go-webauthn/webauthn/metadata"
)

var mds3 metadata.Provider

func ExampleLogic(aaguid string) (err error)
    aaguid, err := uuid.Parse(rawAAGUID)
    if err != nil {
        return err
    }
    
    entry, err := mds3.GetEntry(context.Background(), aaguid)
    if err != nil {
        return err
    }
    
    fmt.Println(entry.MetadataStatement.TcDisplayPNGCharacteristics)
}
```

**Reply by zfLQ2qx2** [reply]
@james-d-elliott Thank you! I was able to use this example to query the data on my Yubikey, and to debug an issue with a device that wasn't in the database - yet, looks like an update will be published next week.

**Reply by james-d-elliott** [reply]
Glad you got this fixed. Also yes, they publish them around the start of the month, monthly.

**Comment by james-d-elliott** [comment]
Also here is a decent few examples of how it's intended to be used (attestation statement validation that's baked into the library, we just parse all of the data for the sake of people may choose to use it for other reasons):

- https://github.com/authelia/authelia/blob/03f06b2df45dffac77ce90a08b19dcfc4ab9a90f/internal/middlewares/authelia_context.go#L702
- https://github.com/authelia/authelia/blob/03f06b2df45dffac77ce90a08b19dcfc4ab9a90f/internal/webauthn/metadata.go
